storage/aptosdb/src/fast_sync_aptos_db.rs

```diff
@@ -0,0 +1,1110 @@
+        *status = FastSyncStatus::FINISHED;
+        Ok(())
+    }
+
+    fn save_transactions(
+        &self,
+        txns_to_commit: &[TransactionToCommit],
+        first_version: Version,
+        base_state_version: Option<Version>,
+        ledger_info_with_sigs: Option<&LedgerInfoWithSignatures>,
+        sync_commit: bool,
+        latest_in_memory_state: StateDelta,
+    ) -> Result<()> {
+        let status = self.get_fast_sync_status()?;
+        if status == FastSyncStatus::STARTED || status == FastSyncStatus::FINISHED {
+            self.db_secondary
+                .as_ref()
+                .expect("db_secondary is None")
+                .save_transactions(
+                    txns_to_commit,
+                    first_version,
+                    base_state_version,
+                    ledger_info_with_sigs,
+                    sync_commit,
+                    latest_in_memory_state,
+                )
+        } else {
+            self.db_main.save_transactions(
+                txns_to_commit,
+                first_version,
+                base_state_version,
+                ledger_info_with_sigs,
+                sync_commit,
+                latest_in_memory_state,
+            )
+        }
+    }
+
+    fn save_transaction_block(
+        &self,
+        txns_to_commit: &[Arc<TransactionToCommit>],
+        first_version: Version,
+        base_state_version: Option<Version>,
+        ledger_info_with_sigs: Option<&LedgerInfoWithSignatures>,
+        sync_commit: bool,
+        latest_in_memory_state: StateDelta,
+        block_state_updates: ShardedStateUpdates,
+        sharded_state_cache: &ShardedStateCache,
+    ) -> Result<()> {
+        let status = self.get_fast_sync_status()?;
+        if status == FastSyncStatus::STARTED || status == FastSyncStatus::FINISHED {
+            self.db_secondary
+                .as_ref()
+                .expect("db_secondary is None")
+                .save_transaction_block(
+                    txns_to_commit,
+                    first_version,
+                    base_state_version,
+                    ledger_info_with_sigs,
+                    sync_commit,
+                    latest_in_memory_state,
+                    block_state_updates,
+                    sharded_state_cache,
+                )
+        } else {
+            self.db_main.save_transaction_block(
+                txns_to_commit,
+                first_version,
+                base_state_version,
+                ledger_info_with_sigs,
+                sync_commit,
+                latest_in_memory_state,
+                block_state_updates,
+                sharded_state_cache,
+            )
+        }
+    }
+}
+
+impl DbReader for FastSyncStorageWrapper {
+    fn get_epoch_ending_ledger_infos(
+        &self,
+        start_epoch: u64,
+        end_epoch: u64,
+    ) -> Result<EpochChangeProof> {
+        if self.is_fast_sync_bootstrap_finished() {
```

Comment:
Eventually, I could imagine us defining a macro for this (if we wanted to clean things up -- especially since all the switch statements are identical 😄 )